keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,382 @@
+
+Pro: 
+- a kind of "implied conduct" from the user's side. The user explicitly defines with the usage of the annotation that the usage of multiple protocols on the same LoadBalancer is accepted
+- Immediate feedback from the K8s API if the user configures a Service with mixed protocol set but without this annotation
+Con:
+- Additional configuration task for the user
+- Must be executed for each and every Service definitions that are to define different protocols for the same LB
+
+#### Merging Services in CPI
+
+This one is not really a classic option control mechanism. The idea comes from the current practice implemented in MetalLB: https://metallb.universe.tf/usage/#ip-address-sharing
+The same works in GCE as a workaround.
+
+I.e. if a cloud provider wants to support this feature the CPI must have a logic to apply the Service definitions with a common key value (for example loadBalancerIP) on the same LoadBalancer instance. If the CPI does not implement this support it will work as it does currently.
+Pro:
+- the cloud provider can decide when to support this feature, and until that it works as currently for this kind of config
+- the restriction on mixed protocols can remain in the K8s API - i.e. there is no K8s API change, and the corresponding risks are mitigated
+Con:
+- the users must maintain two Service instances
+- Atomic update can be a problem - the key must be such a Service attribute that cannot be patched
+
+#### Proposed solution
+
+In the first release: 
+ - a feature flag shall control whether loadbalancer Services with mixed protcol configuration are accepted or not at validation time
+ - we must add a note to the documentation that if such Service is created then it may break things after a rollback - it depends on the cloud provider implementation
+ - the update of such services shall be possible even if the feature flag is OFF. This is to prepare for the next release when the feature flag is removed from the create path, too, and after a rollback to the first release the update of existing Service objects must be possible
+ - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. It does not mean that the CPI implementations have to accept such Services. It means that the CPI implementations shall be able to process such Services. The CPI implementation can still reject such a Service if it cannot manage the corresponding cloud load balancers for those Services, or for any other reasons. The point is, that the system (Service - CPI implementation - cloud LB) must be in a consitent state in the end, and the user must have meaningful information about that state. For exempla, if the Service is not accepted by the CPI implementation this fact shall be available for the user.
```

Comment:
What does it mean to "reject" a service?

Reply:
Well, just like currently some CPIs do not accept a service if for example not all ports are TCP.  My assumption was that that is a kind of accepted way how a CPI can "reject" a service. Similarly, the CPI can still "reject" the service on the same way if it does not want to support this feature.

Reply:
Reworded.